linode/ansible_linode
# new: Add modules for Placement group

## 📝 Description

This PR includes the following change for VM placement functionalities:
1. Add new modules:
    - placement_group: accept both label and id as identifier
    - placement_group_info
    - placement_group_list

2. In the `instance` module, allow to create a new linode instance under a placement group.

## ✔️ How to Test

Run `pip install --force -r requirements.txt`  to install required dependencies. 

Also, you need to set up environment variables to test against alpha:

```
export TEST_API_CA=$PWD/cacert.pem
export TEST_API_URL=https://.../
export LINODE_TOKEN=...
```

**Integration tests:**
```
make TEST_ARGS="-v placement_group" test
make TEST_ARGS="-v placement_group_info" test
make TEST_ARGS="-v placement_group_list" test
make TEST_ARGS="-v instance_basic" test
```

**Manual Test:**

1. In a sandbox environment, i.e. dx-devenv, run the following ansible playbook
```yaml
- name: ansible_linode Sandbox Playbook
  hosts: localhost
  tasks:
    - name: Create a Linode placement group
      linode.cloud.placement_group:
        label: my-ansible-test-pg
        region: us-east
        affinity_type: anti_affinity:local
        state: present
      register: pg_created

    - name: Get a Linode placement group info
      linode.cloud.placement_group_info:
        id: '{{ pg_created.placement_group.id }}'
      register: pg

    - name: Update a Linode placement group label
      linode.cloud.placement_group:
        id: '{{ pg.placement_group.id }}'
        label: '{{ pg.placement_group.label }}-updated'
        state: present
      register: pg_updated

    - name: List Linode placement groups for the current account
      linode.cloud.placement_group_list:
        filters:
          - name: label
            values: ['{{ pg_updated.placement_group.label }}']
      register: pg_list
```

2. Observe no error is raised.

3. Create an instance under the placement group
```yaml
    - name: Create a Linode instance under a placement group
      linode.cloud.instance:
        label: 'ansible-test-pg-{{ r }}'
        type: g6-nanode-1
        region: us-east
        placement_group:
          id: '{{ pg_updated.placement_group.id }}'
          compliant_only: false
        state: present
      register: instance_pg
```

4. Observe that the instance is created successfully.

5. Clean up the resources created.
```yaml
    - name: Delete a Linode instance
      linode.cloud.instance:
        label: '{{ instance_pg.instance.label }}'
        state: absent
      register: delete_instance_pg

    - name: Delete a placement group
      linode.cloud.placement_group:
        label: '{{ pg_updated.placement_group.label }}'
        state: absent
      register: pg_deleted
```
